### PR TITLE
Initial commit for hyp test

### PR DESCRIPTION
**Pull request by sanath-upadhya**
Have added the hypothesis testing framework for Pythoness

### CODE REVIEW

File: pythoness/pythoness_module.py [hunk 1]
```diff
@@ -122,16 +125,18 @@ def close(self):
         self.connection.close()
 
 
-def complete(user_prompt):
+def complete(user_prompt,history):
```
Comment (emeryberger) [inline pythoness/pythoness_module.py]: Is `history` a placeholder for future chat interactions? Also, please add types and docstrings.

File: pythoness/pythoness_module.py [hunk 2]
```diff
@@ -273,7 +286,10 @@ def wrapper(*args, **kwargs):
 
                 # Retry until success.
                 if not function_def:
-                    result = complete(prompt)
+                    result = complete(prompt, history)
+                    # print("History is ")
```
Comment (emeryberger) [inline pythoness/pythoness_module.py]: Remove comment cruft.

File: pythoness/pythoness_module.py [hunk 3]
```diff
@@ -332,7 +348,31 @@ def wrapper(*args, **kwargs):
...
                         try:
-                            if not eval(t):
+                            if isinstance(t, tuple):
+                                compiled_hypothesis_test = create_hypothesis_test(t)
+                                exec(compiled_hypothesis_test,globals())
+                            else:
+                                if not eval(t):
+                                    failing_tests.add(t)
+                        except AssertionError:
+                            #print(locals())
```
Comment (emeryberger) [inline pythoness/pythoness_module.py]: Remove comment

File: pythoness/pythoness_module.py [hunk 4]
```diff
@@ -332,7 +348,31 @@ def wrapper(*args, **kwargs):
...
+                            exc_type, exc_value, exc_tb = sys.exc_info()
+                            tb = traceback.TracebackException(exc_type, exc_value, exc_tb)
+                            exception_info = tb.format_exception_only()
+                            line_number = 0
+                            falsifying_example = get_falsifying_example(exception_info)
+                            for exception_line in exception_info:
+                                print(str(line_number) + " "+ exception_line)
+                                line_number += 1
+
+                            print("Falsifying example is "+ falsifying_example)
```
Comment (emeryberger) [inline pythoness/pythoness_module.py:369]: This is for debugging purposes, right? If so, guard with a debug flag or use the `logging` library and make it a log.debug message.

File: pythoness/pythoness_module.py [hunk 5]
```diff
@@ -332,7 +348,31 @@ def wrapper(*args, **kwargs):
...
+                            if isinstance(t,tuple):
+                                new_l = list(t)
```
Comment (emeryberger) [inline pythoness/pythoness_module.py:373]: Add a comment to explain what's going on here

File: pythoness/pythoness_module.py [hunk 6]
```diff
@@ -398,6 +438,44 @@ def wrapper(*args, **kwargs):
 
     return decorator
 
+def get_falsifying_example(exception_info):
```
Comment (emeryberger) [inline pythoness/pythoness_module.py]: Add types, docstrings

File: test/test-fib.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import pythoness
 
-@pythoness.spec("Compute the nth number in the Fibonacci series.")
+@pythoness.spec("Compute the nth number in the Fibonacci series.",tests = [({'n':'integers(1,20)'}, "myfib(n+3) == myfib(n+1)+myfib(n)")])
```
Comment (emeryberger) [inline test/test-fib.py]: Isn't this hypothesis test wrong? fib(n+3) = fib(n + 2) + fib(n + 1)...

**Review by emeryberger — CHANGES_REQUESTED** [review]

File: pythoness/pythoness_module.py (hunk 1 above)
Comment (jaltmayerpizzorno) [reply pythoness/pythoness_module.py]: Pls also add space after comma.

**Comment by emeryberger** [comment]
Bump @sanath-upadhya

File: pythoness/pythoness_module.py (hunk 1 above)
Comment (sanath-upadhya) [reply pythoness/pythoness_module.py]: Yes, this is a placeholder for keeping track of the conversation which will be used in the future. Added the types and docstrings and updating the pull request now.

**Pull request merged**